pkg/kds/service/server.go

```diff
@@ -136,26 +164,46 @@ func (g *GlobalKDSServiceServer) streamEnvoyAdminRPC(
 		return status.Error(codes.Internal, "could not store stream connection")
 	}
 	logger.Info("stored stream connection")
-	for {
-		resp, err := recv()
-		if err == io.EOF {
-			logger.Info("stream stopped")
-			return nil
+	streamResult := make(chan error, 2)
+	streamReadEnded := make(chan struct{})
+	go func() {
+		select {
+		case <-shouldDisconnectStream.Recv():
+			streamResult <- nil
+		case <-streamReadEnded:
+			return
 		}
-		if status.Code(err) == codes.Canceled {
-			logger.Info("stream cancelled")
-			return nil
-		}
-		if err != nil {
-			logger.Error(err, "could not receive a message")
-			return status.Error(codes.Internal, "could not receive a message")
-		}
-		logger.V(1).Info("Envoy Admin RPC response received", "requestId", resp.GetRequestId())
-		if err := rpc.ResponseReceived(clientID, resp); err != nil {
-			logger.Error(err, "could not mark the response as received")
-			return status.Error(codes.InvalidArgument, "could not mark the response as received")
+	}()
+	go func() {
+		defer func() {
+			close(streamReadEnded)
+		}()
+		for {
+			resp, err := recv()
```

Comment:
Is it guaranteed if we return from `streamEnvoyAdminRPC` func (because of the message in `shouldDisconnectStream`) we won't block the goroutine on `recv()`?

Reply:
The assumption is that the stream ends, so we should get EOF or an error here.

Reply:
From my testing locally we always get `codes.Canceled` and hit the second `if`.